File: src/main/java/org/academy/api/client/gui/command/PosColorRectDrawCommand.java
package org.academy.api.client.gui.command;

import com.mojang.blaze3d.pipeline.RenderPipeline;
import com.mojang.blaze3d.vertex.VertexConsumer;
import org.academy.api.client.render.TextureBinding;
import org.academy.api.client.render.UniformPayload;
import org.joml.Matrix4f;

import java.util.List;

public abstract class PosColorRectDrawCommand extends DrawCommand {
    protected final float width;
    protected final float height;
    protected final float red;
    protected final float green;
    protected final float blue;
    protected final float alpha;

    protected PosColorRectDrawCommand(
            RenderPipeline pipeline,
            float width,
            float height,
            float red,
            float green,
            float blue,
            float alpha,
            List<TextureBinding> textures,
            List<UniformPayload<?>> uniforms
    ) {
        super(pipeline, textures, uniforms);
        this.width = width;
        this.height = height;
        this.red = red;
        this.green = green;
        this.blue = blue;
        this.alpha = alpha;
    }

    @Override
    public void generateVertices(VertexConsumer consumer, Matrix4f pose) {
        consumer.addVertex(pose, 0.0F, 0.0F, 0.0F).setColor(red, green, blue, alpha);
        consumer.addVertex(pose, 0.0F, height, 0.0F).setColor(red, green, blue, alpha);
        consumer.addVertex(pose, width, height, 0.0F).setColor(red, green, blue, alpha);
        consumer.addVertex(pose, width, 0.0F, 0.0F).setColor(red, green, blue, alpha);
    }
}

File: src/main/kotlin/org/academy/api/client/gui/command/PosColorRectDrawCommand.kt
package org.academy.api.client.gui.command

import com.mojang.blaze3d.pipeline.RenderPipeline
import com.mojang.blaze3d.vertex.PoseStack
import org.academy.api.client.gui.render.VertexWriter
import org.academy.api.client.render.TextureBinding
import org.academy.api.client.render.UniformPayload
import org.joml.Vector3f

abstract class PosColorRectDrawCommand protected constructor(
    pipeline: RenderPipeline,
    protected val width: Float,
    protected val height: Float,
    protected val red: Float,
    protected val green: Float,
    protected val blue: Float,
    protected val alpha: Float,
    textures: List<TextureBinding>,
    uniforms: List<UniformPayload<*>>
) : DrawCommand(pipeline, textures, uniforms) {
    override fun generateVertices(writer: VertexWriter, pose: PoseStack.Pose, alphaMul: Float) {
        val matrix = pose.pose()
        val r = (red * 255.0f).toInt()
        val g = (green * 255.0f).toInt()
        val b = (blue * 255.0f).toInt()
        val a = (alpha * alphaMul * 255.0f).toInt()
        val dest = Vector3f()

        writer.beginVertex()
        matrix.transformPosition(0.0f, 0.0f, 0.0f, dest)
        writer.putVec3f(dest.x, dest.y, dest.z)
        writer.putColor(r, g, b, a)

        writer.beginVertex()
        matrix.transformPosition(0.0f, height, 0.0f, dest)
        writer.putVec3f(dest.x, dest.y, dest.z)
        writer.putColor(r, g, b, a)

        writer.beginVertex()
        matrix.transformPosition(width, height, 0.0f, dest)
        writer.putVec3f(dest.x, dest.y, dest.z)
        writer.putColor(r, g, b, a)

        writer.beginVertex()
        matrix.transformPosition(width, 0.0f, 0.0f, dest)
        writer.putVec3f(dest.x, dest.y, dest.z)
        writer.putColor(r, g, b, a)
    }
}
